tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/ImportContextImpl.java

```diff
@@ -66,13 +70,20 @@ public String importType(String typeExpression) {
 
 		// strip off type annotations and '? super' or '? extends'
 		String preamble = "";
-		if ( result.startsWith("@") || result.startsWith("?") ) {
+		if ( result.startsWith( "@" ) ) {
 			int index = result.lastIndexOf(' ');
 			if ( index > 0 ) {
 				preamble = result.substring( 0, index+1 );
 				result = result.substring( index+1 );
 			}
 		}
+		else if ( result.startsWith( "?" ) ) {
+			final Matcher m = P_IMPORT.matcher( result );
```

Comment:
I would much prefer not to see a regex here. 

- It should be easy enough to match `super` or `extends` using `String.indexOf`. 
- And then to match the bound (which may itself be generic type, right?), I would have expected you would need a recursive function. 

[Regexes can't match brackets, so they're not usually appropriate for use in parsing tasks (as opposed to scanning/lexing tasks). This looks more like a parser to me.]

Reply:
Added recursive importStyle invocation and replaced regular expression with equivalent String operations.